Add Benchmark concepts of service time and latency (#5916)
opensearch-project/documentation-website: _benchmark/user-guide/concepts.md

- One or more data streams that are ingested into indexes.
- A set of queries and operations that are invoked as part of the benchmark.

## Throughput and latency

At the end of each test, OSB produces a table that summarizes the following 

- [Service time](#service-time) 
- Throughput
- [Latency](#latency)
- The error rate for each task or OpenSearch operation that ran 

> While the definition for _throughput_ remains consistent with other client-server systems, the definitions for `service time` and `latency` differ from most client-server systems in the context of OSB. The following table compares how service time and latency work with OSB versus the common definition of each term for a client-server system.

Reviewer: Could we also add a sentence at the end of this one along the lines of "For more in-depth explanations of Service Time and Latency, see the following sections on Service Time and Latency"

> While the definition for _throughput_ remains consistent with other client-server systems, the definitions for `service time` and `latency` differ from most client-server systems in the context of OSB. The following table compares how service time and latency work with OSB versus the common definition of each term for a client-server system.

Reviewer 2: ```suggestion
While the definition for _throughput_ remains consistent with other client-server systems, the definitions for `service time` and `latency` differ from most client-server systems in the context of OSB. The following table compares the OSB definition of service time and latency versus the common definitions for a client-server system.
```

| Metric | Common definition | **OSB Definition**	|
| :--- | :--- |:--- |

> | **Throughput** | The number of operations completed in a given period of time	| The number of operations completed in a given period of time	|

Reviewer 2: ```suggestion
| **Throughput** | The number of operations completed in a given period of time.	| The number of operations completed in a given period of time. |
```

> | **Service time**	| The time the server takes to process a request, from the point it receives the request to the point the response is returned. </br></br> It includes the time spent waiting in server-side queues, but _excludes_ network latency, load-balancer overhead, and deserialization/serialization. | The time it takes for the `opensearch-py` to send a request to and receive a response from the OpenSearch cluster. </br> </br> It includes the time it takes for the server to process a request, and also _includes_ network latency, load-balancer overhead, and deserialization/serialization.  |

Reviewer 2: ```suggestion
| **Service time**	| The time the server takes to process a request, from the point it receives the request to the point the response is returned. </br></br> It includes the time spent waiting in server-side queues, but _excludes_ network latency, load-balancer overhead, and deserialization/serialization. | The time it takes for `opensearch-py` to send a request and receive a response from the OpenSearch cluster. </br> </br> It includes the time it takes for the server to process a request, and also _includes_ network latency, load-balancer overhead, and deserialization/serialization.  |
```

> | **Service time**	| The amount of time that the server takes to process a request, from the point it receives the request to the point the response is returned. </br></br> It includes the time spent waiting in server-side queues but _excludes_ network latency, load balancer overhead, and deserialization/serialization. | The amount of time that it takes for `opensearch-py` to send a request and receive a response from the OpenSearch cluster. </br> </br> It includes the amount of time that it takes for the server to process a request and also _includes_ network latency, load balancer overhead, and deserialization/serialization.  |

Reviewer 3: Correct me if I am wrong @dblock, @IanHoang.

In OpenSearch benchmarks, Service Time is calculated in osbenchmark/client.py using the following trace configuration:

```
trace_config = aiohttp.TraceConfig()
trace_config.on_request_start.append(on_request_start)
trace_config.on_request_end.append(on_request_end)
```
I believe the first definition (common definition) in the above documentation code aligns more accurately with our interpretation of Service Time.

Service Time: Represents the interval from the server receiving the request to the server sending the response.

Additional Information:
I attempted to calculate the times by implementing a function perform_request within the AIOHttpConnection in osbenchmark/async_connection.py. The times obtained indicate that the calculated service time doesn't include client processing time.

```
async def perform_request(self, method, url, params=None, body=None, timeout=None, ignore=(), headers=None):
    print("AIOHttpConnection perform_request start time", time.perf_counter())
    status, headers, raw_data = await super().perform_request(method=method, url=url, params=params, body=body, timeout=timeout, ignore=ignore, headers=self.headers)
    print("AIOHttpConnection perform_request end time", time.perf_counter())
    return status, headers, raw_data
```
Sample output:
```
AIOHttpConnection perform_request start time 19.0016295
AIOHttpConnection perform_request end time 19.007145792
service time start 19.001905542
service time end 19.007082167
service time 0.005176625000000712
AIOHttpConnection perform_request start time 19.008452584
AIOHttpConnection perform_request end time 19.012769334
service time start 19.008625875
service time end 19.012713375
service time 0.004087500000000688
```

> | **Latency** | The total time, including the service time and the time the request waited before responding. | The total the request waited before responding based on the `target-throughput` set and any other delays that occur before the request responds. |

Reviewer: `The total the request waited before responding based on the "target-throughput" set and any other delays that occur before the request responds.` -> `The total time the request waited before receiving the response in addition to any other delays that occur before the request is sent. Is based on target-throughput that is set by the user.`

> | **Latency** | The total time, including the service time and the time the request waited before responding. | The total the request waited before responding based on the `target-throughput` set and any other delays that occur before the request responds. |

Author: ```suggestion
| **Latency** | The total time, including the service time and the time the request waited before responding. | Based on the `target-throughput` set by the user, the total time the request waited before receiving the response, in addition to any other delays that occur before the request is sent. |
```

### Service time

OSB does not have insight into how long OpenSearch takes to process a request, apart from extracting the `took` time for the request. In OpenSearch, **service time** tracks the durations between the point OpenSearch issues a requests and when OpenSearch receives the response.

> OSB makes function calls to `opensearch-py`, a client for OpenSearch, to communicate with an OpenSearch cluster. OSB tracks the time between when the `opensearch-py` client sends the request and receives a response from the OpenSearch cluster and considers this to be the service time. Contrasting with the traditional definition of service time, OSB’s definition of service time includes overhead, such as network latency, load-balancer overhead, or deserialization/serialization. The following image illustrates how OpenSearch measure service time versus the traditional measurement. 

Reviewer: `measure` -> `measures`

Reviewer: Also let's rephrase `The following image illustrates how OpenSearch measure service time versus the traditional measurement.`
to
`The following image highlights the difference between traditional service time and OSB's view of service time`

> OSB makes function calls to `opensearch-py`, a client for OpenSearch, to communicate with an OpenSearch cluster. OSB tracks the time between when the `opensearch-py` client sends the request and receives a response from the OpenSearch cluster and considers this to be the service time. Contrasting with the traditional definition of service time, OSB’s definition of service time includes overhead, such as network latency, load-balancer overhead, or deserialization/serialization. The following image illustrates how OpenSearch measure service time versus the traditional measurement. 

Author: ```suggestion
OSB makes function calls to `opensearch-py` to communicate with an OpenSearch cluster. OSB tracks the time between when the `opensearch-py` client sends the request and receives a response from the OpenSearch cluster and considers this to be the service time. Unlike the traditional definition of service time, OSB’s definition of service time includes overhead, such as network latency, load-balancer overhead, or deserialization/serialization. The following image highlights the differences in the traditional definition of service time and OSB's definition of service time. 
```

OSB makes function calls to `opensearch-py` to communicate with an OpenSearch cluster. OSB tracks the time between when the `opensearch-py` client sends the request and receives a response from the OpenSearch cluster and considers this to be the service time. Unlike the traditional definition of service time, OSB’s definition of service time includes overhead, such as network latency, load-balancer overhead, or deserialization/serialization. The following image highlights the differences in the traditional definition of service time and OSB's definition of service time.

<img src="{{site.url}}{{site.baseurl}}/images/benchmark/service-time.png" alt="">

### Latency

> Target throughput is core to understanding OSB’s definition of **latency**. Target throughput is the rate at which OSB issues requests  and assumes the responses will come back instantaneously.  `target-throughput` is one of the common workload parameters that can be set for each test and is measured in operations per second.

Reviewer 2: ```suggestion
Target throughput is core to understanding OSB’s definition of **latency**. Target throughput is the rate at which OSB issues requests and assumes the responses will come back instantaneously.  `target-throughput` is one of the common workload parameters that can be set for each test and is measured in operations per second.
```

> OSB always issue requests one at a time for a single client thread, specified as `search-clients` in workload parameters. If `target-throughput` is set to `0`, OSB issues a request immediately after it receives the response from the previous request. If the `target-throughput` is not 0, OSB issues the next request to match the `target-throughput` set, assuming responses return instantaneously.

Reviewer 2: ```suggestion
OSB always issue requests one at a time for a single client thread, specified as `search-clients` in workload parameters. If `target-throughput` is set to `0`, OSB issues a request immediately after it receives the response from the previous request. If the `target-throughput` is not 0, OSB issues the next request to match the `target-throughput` set, assuming responses return instantaneously.
```
```suggestion
OSB always issues one request at a time for a single client thread, specified as `search-clients` in the workload parameters. If `target-throughput` is set to `0`, OSB issues a request immediately after it receives the response from the previous request. If the `target-throughput` is not 0, OSB issues the next request to match the `target-throughput` set, assuming responses return instantaneously.
```

#### Example A

The following diagrams illustrate how latency is calculated when we expect a request response time of 200ms and use the following settings: 

- `search-clients` is set to `1` 

> - `target-throughput` is set to `1` operations per second. 

Reviewer 2: ```suggestion
- `target-throughput` is set to `1` operations per second
```

However, when a request takes longer than 200ms, such as when a request takes 1110ms instead of 400ms, OSB sends the next request that was supposed to occur at 4.00s based on the `target-throughput` at 4.10s. All subsequent requests after the 4.10s request attempt to resynchronize with the `target-throughput` setting.

<img src="{{site.url}}{{site.baseurl}}/images/benchmark/latency-explanation-2.png" alt="">

When measuring the overall latency, OSB looks all the requests performed. All requests have a latency of 200ms except for the following two: 
- The request that lasted 1100ms 
- The following request which was supposed to start at 4:00s. This request was delayed by 100ms, denoted by the orange area in the following diagram, and had a response of 200ms. 

<img src="{{site.url}}{{site.baseurl}}/images/benchmark/latency-explanation-3.png" alt="">

> Therefore, the overall latency OSB calculates in this scenario, when accounting for the delayed response time with the response time, is **300ms**.

Reviewer: A couple of things: 
1. Could we move this sentence up and join line 75 so that readers know that it relates to this request and not all requests. 

2. Could we rephrase this line from `the overall latency OSB calculates in this scenario, when accounting for the delayed response time with the response time, is **300ms**.` to `When calculating the latency for this request, OSB will account for the delayed start time and combine it with the response time. Thus, the latency for this request is **300ms**.`

Author: Is the latency measurement specific to the request or the average latency for all requests @IanHoang?

Reviewer: it's specific to the request

#### Example B

> In this example, OSB uses the following latency settings:

Reviewer: instead of `OSB uses the following latency settings`, it might be more apt to say `the following assumptions are made:`

Author: Only one of these bullets is an assumption. We could say "In this example, OSB assumes a request response time of 200ms, and uses the following settings:"

Thoughts @IanHoang?

Reviewer: Yes that works better 👍🏻

> - Assumes a 200ms response time

Reviewer: `Assumes a 200ms response time` -> `Assumes the request's response time is always 200ms`

> - `search_clients` set to 1

Reviewer 4: Should "1" be in code font?

> - `target-throughput` set to 10 operations per second

Reviewer 4: ```suggestion
- `target-throughput` is set to 10 operations per second.
```

> - `target-throughput` set to 10 operations per second

Reviewer 4: Should "10" be in code font?

> The following diagram shows the schedule built by OSB with the expected response time.

Reviewer 4: ```suggestion
The following diagram shows the schedule built by OpenSearch Benchmark with the expected response time.
```

> The following diagram shows the schedule built by OSB with the expected response time.

Reviewer 4: Should "time" be "times"?

<img src="{{site.url}}{{site.baseurl}}/images/benchmark/b-latency-explanation-1.png" alt="">

> However, if the assumption is that all responses will take 200ms, 10 operations per second won't be possible. Therefore, the highest throughput OSB can reach is 5 operations per second, as shown in the following diagram.

Reviewer 4: ```suggestion
However, if the assumption is that all responses will take 200ms, 10 operations per second won't be possible. Therefore, the highest throughput OpenSearch Benchmark can reach is 5 operations per second, as shown in the following diagram.
```

<img src="{{site.url}}{{site.baseurl}}/images/benchmark/b-latency-explanation-2.png" alt="">

> OSB does not account for this, and continues to try to achieve the `target-throughput` of 10 operations per second. Because of this, delays for each request being to cascade, as illustrated in the following diagram.

Reviewer 4: ```suggestion
OpenSearch Benchmark does not account for this and continues to try to achieve the `target-throughput` of 10 operations per second. Because of this, delays for each request begin to cascade, as illustrated in the following diagram.
```

<img src="{{site.url}}{{site.baseurl}}/images/benchmark/b-latency-explanation-3.png" alt="">

> If we combine the service time with the delay for each operation, we’ll get the following latency measurements for each operation: 

Reviewer 4: ```suggestion
Combining the service time with the delay for each operation provides the following latency measurements for each operation: 
```

- 200 ms for operation 1
- 300 ms for operation 2
- 400 ms for operation 3
- 500 ms for operation 4 
- 600 ms for operation 5

This latency cascade continues, increasing latency by 100ms for each subsequent request.

### Recommendation

> As shown in the preceding examples, be aware of the average service time of each task run and provide a `target-throughput` that accounts for the service time. OSB’s latency is calculated based on the `target-throughput` set by the user. In other words, OSB’s latency could be redefined as "throughput-based latency".

Reviewer 4: ```suggestion
As shown by the preceding examples, you should be aware of the average service time of each task and provide a `target-throughput` that accounts for the service time. The OpenSearch Benchmark latency is calculated based on the `target-throughput` set by the user. In other words, the OpenSearch Benchmark latency could be redefined as "throughput-based latency".
```
